Adding unit tests for the NoDL conversion API

aprotyas opened the pull request:
This PR adds unit tests for the `nodl_to_policy.{policy, _common._profile}` modules, as well as ancillary test fixtures.

Also:
* Fixed some implementation details in the `policy` module.
* Added some type hinting in the `_common/_profile` module.
* Added `sros` as an explicit dependency in the `package.xml` manifest.
* Improved organization of unit test fixtures.

It's a big PR, but a lot of it is relatively simple additions.

Signed-off-by: Abrar Rahman Protyasha <abrar@openrobotics.org>

Commit:
Adding unit tests for the NoDL conversion API

Also:
* Fixed some implementation details in the `policy` module.
* Added some type hinting in the `_common/_profile` module.
* Added `sros` as an explicit dependency in the `package.xml` manifest.
* Improved organization of unit test fixtures.

Signed-off-by: Abrar Rahman Protyasha <abrar@openrobotics.org>

aprotyas commented on nodl_to_policy/_common/_profile.py:
```diff
@@ -22,7 +22,7 @@
 
 # TODO(aprotyas): For now, only use `node.xml`, since NoDL does not describe Lifecycle nodes
-def common_profile() -> etree.ElementTree:
+def common_profile() -> etree._ElementTree:
```
FYI: Return type changed to `etree._ElementTree` for uniformity throughout source code. Afterall, `etree.ElementTree` is simply a wrapper class around the former.

aprotyas commented:
Pinging @clalancette for a review.

clalancette commented on nodl_to_policy/policy.py:
```diff
@@ -81,13 +81,14 @@ def get_profile(policy: etree._ElementTree, node_name: str) -> etree._ElementTre
         # namespace information not provided in NoDL description yet
         profile.attrib['ns'] = '/'
         profile.attrib['node'] = node_name
+        profile.attrib['ns'] = '/'
```
Isn't this redundant with 2 lines above?

aprotyas replied:
Yes, this line is redundant. Fixed in b8a958c.

clalancette commented on nodl_to_policy/policy.py:
```diff
@@ -146,12 +147,14 @@ def add_permission(
 def add_common_permissions(profile: etree._ElementTree, node: Node) -> None:
     """
-    `add_permission` but for each of the common services/topics/actions.
+    `add_permissions` but for each of the common services/topics/actions.
```
```suggestion
    `add_permissions` for each of the common services/topics/actions.
```

aprotyas replied:
Fixed in a0e5cb6.

clalancette commented on package.xml:
```diff
@@ -15,6 +15,7 @@
   <exec_depend>ros2nodl</exec_depend>
   <exec_depend>ros2pkg</exec_depend>
```
Does this package actually require `ros2pkg` in any way?

aprotyas replied:
Actually, the package doesn't require `ros2pkg`. This should also probably be part of some draft branch where `nodl_to_policy` can be supplied a package name instead of just a NoDL XML file. I would've used the `ros2pkg.api.package_name_completer` to help implement that feature.

Regardless - I've removed this dependency in 74c9fcc.

clalancette commented on test/nodl_to_policy/test_policy.py:
```diff
@@ -0,0 +1,330 @@
+    # check that a single profile tag was created
+    assert len(test_profiles) == 1
+    assert test_profiles[0] == test_profile
+    test_profile.tag == 'profile'
+    test_profile.attrib['ns'] == '/'
+    test_profile.attrib['node'] == 'foo'
```
These are currently "dead" statements; they have no affect.  Should these be `assert`ed?

aprotyas replied:
Yes, these should have been assert statements. Fixed in b8a958c.

clalancette commented on test/nodl_to_policy/test_policy.py:
```diff
@@ -0,0 +1,330 @@
+
+    # check that a <profile node='node1'> tag exists as expected
+    assert test_profiles.find(path=f'profile[@ns="/"][@node="node_1"]') == test_profile
+    test_profile.tag == 'profile'
+    test_profile.attrib['ns'] == '/'
+    test_profile.attrib['node'] == 'node_1'
```
Same question about these dead statements; should they be `assert`ed?

aprotyas replied:
These should have been asserted as well. Fixed in b8a958c.

clalancette reviewed (changes requested):
I've got a few mostly minor things to fix inline.

Commit:
Fixing 'dead'/redundant statements

Signed-off-by: Abrar Rahman Protyasha <abrar@openrobotics.org>

Commit:
Removing unnecessary exec dependency on `ros2pkg`

This dependency is only required for the possible use
of `ros2pkg.api.{package_name_completer, PackageNotFound}`
in case an enhancement to `nodl_to_policy` allows passing
in package names rather than NoDL XML file paths only.

Signed-off-by: Abrar Rahman Protyasha <abrar@openrobotics.org>

Commit:
Fix `add_common_permissions` function description

Signed-off-by: Abrar Rahman Protyasha <abrar@openrobotics.org>

clalancette reviewed (changes requested):
OK, so the code that is in here is looking pretty good to me.  There are 3 things I think need to happen:

1. There is a rogue trailing whitespace in the package.xml that should be removed.
2. When running the tests locally, I saw about 11 flake8 warnings, mostly in the tests.
3. I'm seeing this warning:
```
  Warning: The 'junit_family' default value will change to 'xunit2' in pytest 6.0.
  Add 'junit_family=xunit1' to your pytest.ini file to keep the current format in future versions of pytest and silence this warning.
```
You can silence that by adding a `pytest.ini` file at the top-level that contains:

```
[pytest]
junit_family=xunit2
```

Once those three things are fixed, I think this is ready to merge.

Commit:
Removing trailing whitespace in 'package.xml'

Signed-off-by: Abrar Rahman Protyasha <abrar@openrobotics.org>

Commit:
Add 'pytest.ini' so local tests don't display warning

Signed-off-by: Abrar Rahman Protyasha <abrar@openrobotics.org>

Commit:
Fixing flake8 warnings

Signed-off-by: Abrar Rahman Protyasha <abrar@openrobotics.org>

aprotyas commented:
Thanks for the review! I don't see any more flake8 warnings on my end. Final check, and then we can merge?
